Bug: [new config system] `ignores` does not work without `files`

### Environment

Node version: 18.16.0
npm version: 9.6.4
Local ESLint version: 8.38.0
Global ESLint version: N/A
Operating System: Ubuntu 22.04 (WSL)

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```js
[
  {
    rules: {
      "no-unused-vars": "error",
    },
  },
  {
    ignores: ["error.js"],
    rules: {
      "no-unused-vars": "warn",
    },
  },
];
```

</details>

`./error.js` and `./warn.js`:

```js
let unusedVar;
```

Then run `eslint .`

### What did you expect to happen?

```shell
{cwd}/error.js
  1:5  error  'unusedVar' is defined but never used  no-unused-vars

{cwd}/warn.js
  1:5  warning  'unusedVar' is defined but never used  no-unused-vars
```

### What actually happened?

```shell
{cwd}/error.js
  1:5  warning  'unusedVar' is defined but never used  no-unused-vars

{cwd}/warn.js
  1:5  warning  'unusedVar' is defined but never used  no-unused-vars
```

### Link to Minimal Reproducible Example

https://github.com/filetvignon/eslint-bug-ignores

### Participation

- [X] I am willing to submit a pull request for this issue.

### Additional comments

I set 2 configuration objects: one with a rule all files and another to change that rule for all files EXCEPT `error.js`.
I omitted the `files` property in these objects; although the documentation recommends using `ignores` with `files`, [it states that this should work](https://eslint.org/docs/latest/use/configure/configuration-files-new#excluding-files-with-ignores) and omitting `files` should be the same as using `files: '**/*'`.

In reality, it works only if I explicitly set `files` in the second configuration object:

```js
[
  {
    rules: {
      "no-unused-vars": "error",
    },
  },
  {
    files: ["**/*"],
    ignores: ["error.js"],
    rules: {
      "no-unused-vars": "warn",
    },
  },
];
```

--- a/package.json
+++ b/package.json
@@ -64,7 +64,7 @@
     "@eslint-community/regexpp": "^4.4.0",
     "@eslint/eslintrc": "^2.0.3",
     "@eslint/js": "8.41.0",
-    "@humanwhocodes/config-array": "^0.11.9",
+    "@humanwhocodes/config-array": "^0.11.10",
     "@humanwhocodes/module-importer": "^1.0.1",
     "@nodelib/fs.walk": "^1.2.8",
     "ajv": "^6.10.0",

--- a/package.json
+++ b/package.json
@@ -64,7 +64,7 @@
     "@eslint-community/regexpp": "^4.4.0",
     "@eslint/eslintrc": "^2.0.3",
     "@eslint/js": "8.41.0",
-    "@humanwhocodes/config-array": "github:humanwhocodes/config-array",
+    "@humanwhocodes/config-array": "^0.11.9",
     "@humanwhocodes/module-importer": "^1.0.1",
     "@nodelib/fs.walk": "^1.2.8",
     "ajv": "^6.10.0",

--- a/package.json
+++ b/package.json
@@ -64,7 +64,7 @@
     "@eslint-community/regexpp": "^4.4.0",
     "@eslint/eslintrc": "^2.0.3",
     "@eslint/js": "8.41.0",
-    "@humanwhocodes/config-array": "^0.11.8",
+    "@humanwhocodes/config-array": "github:humanwhocodes/config-array",
     "@humanwhocodes/module-importer": "^1.0.1",
     "@nodelib/fs.walk": "^1.2.8",
     "ajv": "^6.10.0",

--- a/tests/lib/eslint/flat-eslint.js
+++ b/tests/lib/eslint/flat-eslint.js
@@ -4631,6 +4631,99 @@ describe("FlatESLint", () => {
         });
     });
 
+    describe("configs with 'ignores' and without 'files'", () => {
+
+        // https://github.com/eslint/eslint/issues/17103
+        describe("config with ignores: ['error.js']", () => {
+            const cwd = getFixturePath("config-with-ignores-without-files");
+            const { prepare, cleanup, getPath } = createCustomTeardown({
+                cwd,
+                files: {
+                    "eslint.config.js": `module.exports = [
+                        {
+                            rules: {
+                                "no-unused-vars": "error",
+                            },
+                        },
+                        {
+                            ignores: ["error.js"],
+                            rules: {
+                                "no-unused-vars": "warn",
+                            },
+                        },
+                      ];`,
+                    "error.js": "let unusedVar;",
+                    "warn.js": "let unusedVar;"
+                }
+            });
+
+            beforeEach(prepare);
+            afterEach(cleanup);
+
+            it("should apply to all files except for 'error.js'", async () => {
+                const engine = new FlatESLint({
+                    cwd
+                });
+
+                const results = await engine.lintFiles("{error,warn}.js");
+
+                assert.strictEqual(results.length, 2);
+
+                const [errorResult, warnResult] = results;
+
+                assert.strictEqual(errorResult.filePath, path.join(getPath(), "error.js"));
+                assert.strictEqual(errorResult.messages.length, 1);
+                assert.strictEqual(errorResult.messages[0].ruleId, "no-unused-vars");
+                assert.strictEqual(errorResult.messages[0].severity, 2);
+
+                assert.strictEqual(warnResult.filePath, path.join(getPath(), "warn.js"));
+                assert.strictEqual(warnResult.messages.length, 1);
+                assert.strictEqual(warnResult.messages[0].ruleId, "no-unused-vars");
+                assert.strictEqual(warnResult.messages[0].severity, 1);
+            });
+        });
+
+        describe("config with ignores: ['**/*.json']", () => {
+            const cwd = getFixturePath("config-with-ignores-without-files");
+            const { prepare, cleanup, getPath } = createCustomTeardown({
+                cwd,
+                files: {
+                    "eslint.config.js": `module.exports = [
+                        {
+                            rules: {
+                                "no-undef": "error",
+                            },
+                        },
+                        {
+                            ignores: ["**/*.json"],
+                            rules: {
+                                "no-unused-vars": "error",
+                            },
+                        },
+                      ];`,
+                    "foo.js": "",
+                    "foo.json": ""
+                }
+            });
+
+            beforeEach(prepare);
+            afterEach(cleanup);
+
+            it("should not add json files as lint targets", async () => {
+                const engine = new FlatESLint({
+                    cwd
+                });
+
+                const results = await engine.lintFiles("foo*");
+
+                // should not lint `foo.json`
+                assert.strictEqual(results.length, 1);
+                assert.strictEqual(results[0].filePath, path.join(getPath(), "foo.js"));
+            });
+        });
+
+    });
+
     describe("with ignores config", () => {
         const root = getFixturePath("cli-engine/ignore-patterns");
 

